### PR TITLE
CORE-127: automatically publish tools when an app is being published

### PR DESCRIPTION
**Pull request by slr71**
My biggest concern here is that there might be some more corner cases that I haven't thought of. If reviewers could keep this in mind while looking over the code, I'd appreciate it.

### CODE REVIEW

File: src/apps/service/apps/de/validation.clj [hunk 1]
```diff
@@ -45,6 +45,18 @@
...
+(defn- remove-owned-tools
+  [username tools]
+  (let [tool-ids  (mapv :id tools)
+        owned-ids (set (keys (perms-client/load-tool-permissions username tool-ids "own")))]
+    (remove (comp owned-ids :id) tools)))
+
+(defn- get-unpublishable-tools
+  [username tools]
+  (let [public-tool-ids (perms-client/get-public-tool-ids)]
+    (->> (remove (comp public-tool-ids :id) tools)
```
Comment (psarando) [inline src/apps/service/apps/de/validation.clj]: 👍

File: src/apps/service/apps/de/validation.clj [hunk 2]
```diff
@@ -45,6 +45,18 @@
...
+         (remove-owned-tools username))))
```
Comment (psarando) [inline src/apps/service/apps/de/validation.clj]: We had also discussed restricting auto-publish only for tools that had a container image that was already in our private registry, so the image name had to start with `docker.cyverse.org/`.

Should we implement this restriction in these service endpoints as well?

Comment (slr71) [reply src/apps/service/apps/de/validation.clj]: Ah, I'd forgotten about that requirement. Thanks for mentioning it. Yeah, that requirement should be also be in the services.

Comment (slr71) [reply src/apps/service/apps/de/validation.clj]: The validation has been added. I think this is ready for another review.

**Comment by slr71** [comment]
Thanks for the review!

**Comment by slr71** [comment]
s/review/reviews/ 😛

**Pull request merged**